src/base/lib/Convex/Class.hs

```diff
@@ -39,71 +41,74 @@ module Convex.Class(
+import           Control.Monad.Trans.Except.Result                  (ResultT)
+import           Convex.Constants                                   (ERA)
+import           Convex.Eras                                        (InAnyBabbageEraOnwards (..))
+import qualified Convex.Eras                                        as Eras
+import           Convex.MonadLog                                    (MonadLog (..),
+                                                                     MonadLogIgnoreT (..),
+                                                                     logInfoS,
+                                                                     logWarn,
+                                                                     logWarnS)
+import           Convex.Utils                                       (posixTimeToSlotUnsafe,
+                                                                     slotToUtcTime)
+import           Convex.Utxos                                       (UtxoSet)
+import           Data.Aeson                                         (FromJSON,
+                                                                     ToJSON)
+import           Data.Bifunctor                                     (Bifunctor (..))
+import           Data.Map                                           (Map)
+import qualified Data.Map                                           as Map
+import           Data.Set                                           (Set)
+import qualified Data.Set                                           as Set
+import           Data.Time.Clock                                    (UTCTime)
+import           GHC.Generics                                       (Generic)
+import           Ouroboros.Consensus.HardFork.Combinator.AcrossEras (EraMismatch)
+import           Ouroboros.Consensus.HardFork.History               (interpretQuery,
+                                                                     slotToSlotLength)
+import qualified Ouroboros.Network.Protocol.LocalStateQuery.Type    as T
+import           Ouroboros.Network.Protocol.LocalTxSubmission.Type  (SubmitResult (..))
+import qualified PlutusLedgerApi.V1                                 as PV1
+import           Prettyprinter                                      (Pretty (..),
+                                                                     (<+>))
+import           Test.QuickCheck.Monadic                            (PropertyM)
 
 {-| Send transactions and resolve tx inputs.
 -}
 class Monad m => MonadBlockchain m where
   -- see note Note [sendTx Failure]
   sendTx                  :: Tx BabbageEra -> m (Either SendTxFailed TxId) -- ^ Submit a transaction to the network
-  utxoByTxIn              :: Set C.TxIn -> m (C.UTxO C.BabbageEra) -- ^ Resolve tx inputs
-  queryProtocolParameters :: m (LedgerProtocolParameters C.BabbageEra) -- ^ Get the protocol parameters
+  utxoByTxIn              :: Set C.TxIn -> m (InAnyBabbageEraOnwards C.UTxO) -- ^ Resolve tx inputs
```

Comment:
I'm not 100% pushing for this, but why not just use `InAnyShelleyBasedEra` from `cardano-api`? Rationale: one less type to maintain in sc-tools, and we _might_ end up generalizing this to something closer to a real node.

Reply:
At the same time, there's going to be a lot of pattern matching, so definitely not a priority especially if this ticket needs to be merged ASAP.